### PR TITLE
[WIP]: Master rebase

### PR DESCRIPTION
**Pull request by buffalu**
Ow.

This will be squashed + force pushed to master, but the upstream in this PR is upstream/master as of june 1.

will need to perform another rebase to catch up again, but would prefer to rebase 1.16 on this to avoid rebase drift.

in the near future will move to auto-rebasing on a nightly/weekly cadence

### CODE REVIEW

File: build_tip_programs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env sh
```
Comment (buffalu) [inline build_tip_programs:1]: can delete this file since we have the programs in fetch spl + bootstrap args

File: bootstrap [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  NDEBUG=1 ./multinode-demo/bootstrap-validator.sh \
+  --wait-for-supermajority 0 \
+  --expected-bank-hash "$BANK_HASH" \
+  --block-engine-address http://127.0.0.1:1003 \
+  --block-engine-auth-service-address http://127.0.0.1:1005 \
+  --relayer-auth-service-address http://127.0.0.1:11226 \
+  --relayer-address http://127.0.0.1:11226 \
+  --rpc-pubsub-enable-block-subscription \
+  --enable-rpc-transaction-history \
+  --tip-payment-program-pubkey 6veFRUKJBNGMR58LEcKn5Bc6MR17WZF4rsgD4Lqq7fsU \
```
Comment (buffalu) [inline bootstrap:23]: change to mainnet addresses. todo: double check running this w/ start to make sure genesis works correctly w/ deployed programs

File: bundle/src/bundle_execution.rs [hunk 3]
```diff
@@ -0,0 +1,1184 @@
...
+    metrics: BundleExecutionMetrics,
+}
+
+impl<'a> LoadAndExecuteBundleOutput<'a> {
+    pub fn executed_ok(&self) -> bool {
+        self.result.is_ok()
+    }
+
+    pub fn result(&self) -> LoadAndExecuteBundleResult<()> {
+        self.result.clone()
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:50]: need to clone?

File: bundle/src/bundle_execution.rs [hunk 4]
```diff
@@ -0,0 +1,1184 @@
...
+
+        debug!(
+            "bundle: {} batch num locks ok: {}",
+            bundle.bundle_id,
+            batch.lock_results().iter().filter(|lr| lr.is_ok()).count()
+        );
+
+        // Ensures that bundle lock results only return either:
+        // Ok(()) | Err(TransactionError::AccountInUse)
+        // If the error isn't one of those, the
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:284]: nit comment

File: bundle/src/bundle_execution.rs [hunk 5]
```diff
@@ -0,0 +1,1184 @@
...
+        || post_execution_accounts.len() != bundle.transactions.len()
+    {
+        return LoadAndExecuteBundleOutput {
+            bundle_transaction_results: vec![],
+            result: Err(LoadAndExecuteBundleError::InvalidPreOrPostAccounts),
+            metrics: BundleExecutionMetrics::default(),
+        };
+    }
+    let mut binding = AccountOverrides::default();
+    let account_overrides = account_overrides.unwrap_or(&mut binding);
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:243]: nit: consolidate this and above into one liner

File: bundle/src/bundle_execution.rs [hunk 6]
```diff
@@ -0,0 +1,1184 @@
...
+            })
+        } else {
+            None
+        }
+    })
+    .collect()
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:478]: smexi

File: bundle/src/bundle_execution.rs [hunk 7]
```diff
@@ -0,0 +1,1184 @@
...
+            return LoadAndExecuteBundleOutput {
+                bundle_transaction_results,
+                metrics,
+                result: Err(LoadAndExecuteBundleError::ProcessingTimeExceeded(
+                    start_time.elapsed(),
+                )),
+            };
+        }
+
+        let chunk_end = min(bundle.transactions.len(), chunk_start.saturating_add(128));
```
Comment (jedleggett) [inline bundle/src/bundle_execution.rs:263]: Is this just to future proof for jumbo bundles?  Doesn't seem like we'll ever have more than 1 chunk, even if we increase bundle size by 20x current limits.

Comment (buffalu) [reply bundle/src/bundle_execution.rs:263]: yeah

File: core/src/banking_stage/consumer.rs [hunk 8]
```diff
@@ -443,20 +453,26 @@ impl Consumer {
...
             txs,
             pre_results
         ));
 
         // Only lock accounts for those transactions are selected for the block;
         // Once accounts are locked, other threads cannot encode transactions that will modify the
-        // same account state
+        // same account state.
+        // BundleAccountLocker is used to prevent race conditions with bundled transactions from bundle stage
+        let bundle_account_locker = self.bundle_account_locker.account_locks();
```
Comment (segfaultdoc) [inline core/src/banking_stage/consumer.rs:465]:
```suggestion
        let account_locks = self.bundle_account_locker.account_locks();
```

File: core/src/broadcast_stage.rs [hunk 9]
```diff
@@ -412,13 +428,21 @@ pub fn broadcast_shreds(
...
-                    .map(|addr| (shred.payload(), addr))
+            shreds.flat_map(move |shred| {
+                let mut addrs = Vec::with_capacity(2);
+                if let Some(peer) = cluster_nodes.get_broadcast_peer(&shred.id()) {
+                    if let Ok(tvu) = peer.tvu(Protocol::UDP) {
+                        addrs.push(tvu);
+                    }
+                }
+                if let Some(shred_receiver_address) = shred_receiver_address {
+                    addrs.push(*shred_receiver_address);
```
Comment (segfaultdoc) [inline core/src/broadcast_stage.rs:439]: maybe minuscule but thoughts on pushing the first?

File: core/src/bundle_stage.rs [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+
+    // currently buffered
+    current_buffered_bundles_count: AtomicU64,
+    current_buffered_packets_count: AtomicU64,
+
+    // buffered due to cost model
+    cost_model_buffered_bundles_count: AtomicU64,
+    cost_model_buffered_packets_count: AtomicU64,
+
+    // number of packets + bundles dropped during insertion
```
Comment (segfaultdoc) [inline core/src/bundle_stage.rs:72]: comment a little confusing, what's `packets + bundles`

File: core/src/bundle_stage.rs [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+        }
+    }
+}
+
+pub struct BundleStage {
+    bundle_thread: JoinHandle<()>,
+}
+
+impl BundleStage {
+    #[allow(clippy::new_ret_no_self)]
```
Comment (segfaultdoc) [inline core/src/bundle_stage.rs:198]: no necessitas

File: core/src/bundle_stage/bundle_consumer.rs [hunk 12]
```diff
@@ -0,0 +1,1514 @@
...
+    // payment program and set the tip receiver to themself.
+    // A bundle is not allowed to touch consensus-related accounts
+    //  - This is to avoid stalling the voting BankingStage threads.
+    pub fn consume_buffered_bundles(
+        &mut self,
+        bank_start: &BankStart,
+        unprocessed_transaction_storage: &mut UnprocessedTransactionStorage,
+        bundle_stage_leader_metrics: &mut BundleStageLeaderMetrics,
+    ) {
+        self.maybe_update_blacklist(bank_start);
```
Comment (segfaultdoc) [inline core/src/bundle_stage/bundle_consumer.rs:126]: should `blacklisted_accounts` be shared b/w banking stage and here?

Comment (segfaultdoc) [reply core/src/bundle_stage/bundle_consumer.rs:126]: ah nvm

File: core/src/unprocessed_transaction_storage.rs [hunk 13]
```diff
@@ -359,13 +421,60 @@ impl UnprocessedTransactionStorage {
...
+    {
+        match self {
+            UnprocessedTransactionStorage::BundleStorage(bundle_storage) => bundle_storage
+                .process_bundles(
+                    bank,
+                    bundle_stage_leader_metrics,
+                    blacklisted_accounts,
+                    processing_function,
+                ),
+            _ => panic!("class does not support processing bundles"),
```
Comment (segfaultdoc) [inline core/src/unprocessed_transaction_storage.rs:461]: can you assert this earlier in the chain?

File: core/src/bundle_stage/bundle_consumer.rs [hunk 14]
```diff
@@ -0,0 +1,1514 @@
...
+        self.reserved_space.tick(&bank_start.working_bank);
+
+        let reached_end_of_slot = unprocessed_transaction_storage.process_bundles(
+            bank_start.working_bank.clone(),
+            bundle_stage_leader_metrics,
+            &self.blacklisted_accounts,
+            |bundles, bundle_stage_leader_metrics| {
+                Self::do_process_bundles(
```
Comment (segfaultdoc) [inline core/src/bundle_stage/bundle_consumer.rs:134]: should we have init programs check around here?

File: core/src/bundle_stage/bundle_consumer.rs [hunk 15]
```diff
@@ -0,0 +1,1514 @@
...
+        debug!(
+            "bundle: {} reserving blockspace for {} transactions",
+            sanitized_bundle.bundle_id,
+            sanitized_bundle.transactions.len()
+        );
+
+        let (
+            (transaction_qos_cost_results, _cost_model_throttled_transactions_count),
+            _cost_model_elapsed_us,
+        ) = measure_us!(Self::reserve_bundle_blockspace(
```
Comment (segfaultdoc) [inline core/src/bundle_stage/bundle_consumer.rs:504]: should either emit the latency metric or not call `measure`

File: bundle/src/bundle_execution.rs [hunk 16]
```diff
@@ -0,0 +1,1184 @@
...
+                metrics,
+                result: Err(LoadAndExecuteBundleError::LockError {
+                    signature: *transaction.signature(),
+                    transaction_error: lock_failure.clone(),
+                }),
+            };
+        }
+
+        let mut pre_balance_info = PreBalanceInfo::default();
+        let (_, _collect_balances_us) = measure_us!({
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:303]: why call measure?

File: bundle/src/bundle_execution.rs [hunk 17]
```diff
@@ -0,0 +1,1184 @@
...
+    let account_overrides = account_overrides.unwrap_or(&mut binding);
+
+    let mut chunk_start = 0;
+    let start_time = Instant::now();
+
+    let mut bundle_transaction_results = vec![];
+    let mut metrics = BundleExecutionMetrics::default();
+
+    while chunk_start != bundle.transactions.len() {
+        if start_time.elapsed() > *max_processing_time {
```
Comment (segfaultdoc) [inline bundle/src/bundle_execution.rs:252]: wondering if it makes sense to have this check in `load_and_execute_transactions` after each tx? assuming we send megabundles in the future might make sense

**Comment by segfaultdoc** [comment]
still going thru, lots of code im familiar with but some other places not

File: bundle/src/bundle_execution.rs (hunk 17 above)
Comment (buffalu) [reply bundle/src/bundle_execution.rs:252]: maybe in the future, for now i think this makes sense. ideally the 128 tx cap and/or bundle size catches us for now

File: bundle/src/bundle_execution.rs (hunk 3 above)
Comment (buffalu) [reply bundle/src/bundle_execution.rs:50]: can't move out of self

Comment (segfaultdoc) [reply bundle/src/bundle_execution.rs:50]: was thinking return a reference and figure it out upstream but minor

File: core/src/unprocessed_transaction_storage.rs (hunk 13 above)
Comment (buffalu) [reply core/src/unprocessed_transaction_storage.rs:461]: wym?

**Pull request closed without merging**